tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla

```diff
@@ -0,0 +1,446 @@
+    /\ dirty' = TRUE
+    /\ length' = Max(length, curr')
+    /\ UNCHANGED <<lo, diskPos, file_pointer, file_content>>
+
+Read(data) ==
+    LET numReadableWithoutSeeking == Min(lo + BuffSz, length) - curr IN
+    /\ Inv2
+    /\ numReadableWithoutSeeking >= 0
+    /\ LET numToRead == Min(ArrayLen(data), numReadableWithoutSeeking) IN
+       LET buffOff == curr - lo IN
+        /\ data = ArraySlice(buff, buffOff, buffOff + numToRead)
+        /\ curr' = curr + numToRead
+    /\ UNCHANGED <<buff, dirty, diskPos, file_content, file_pointer, length, lo>>
+
+\* The `write()` method is composed of repeated calls to `writeAtMost()`, so
+\* verifying that the latter maintains all our invariants should be sufficient.
+WriteAtMost(data) ==
+    LET numWriteableWithoutSeeking == Min(ArrayLen(data), lo + BuffSz - curr) IN
+    LET buffOff == curr - lo IN
+    /\ Inv2
+    /\ curr + numWriteableWithoutSeeking <= MaxOffset
+    /\ buff' = ArrayConcat(ArrayConcat(
+            ArraySlice(buff, 0, buffOff),
+            ArraySlice(data, 0, numWriteableWithoutSeeking)),
+            ArraySlice(buff, buffOff + numWriteableWithoutSeeking, ArrayLen(buff)))
+    /\ dirty' = TRUE
+    /\ curr' = curr + numWriteableWithoutSeeking
+    /\ length' = Max(length, curr')
+    /\ UNCHANGED <<lo, diskPos, file_content, file_pointer>>
+
+Next ==
+    \/ FlushBuffer
+    \/ \E offset \in Offset:
+        \/ Seek(offset)
+        \/ SetLength(offset)
+    \/ \E symbol \in Symbols:
+        \/ Read1(symbol)
+        \/ Write1(symbol)
+    \/ \E len \in 1..MaxOffset: \E data \in Array(Symbols, len):
+        \/ WriteAtMost(data)
+        \/ Read(data)
+
+Spec == Init /\ [][Next]_vars
+
+-------------------------------------------------------------------------------
+\* Refinement of general RandomAccessFile
+
+RAF == INSTANCE RandomAccessFile WITH
+    file_content <- LogicalFileContent,
+    file_pointer <- curr
+
+Safety == RAF!Spec
+
+\* Ensure that the various actions behave according to their abstract specifications.
+FlushBufferCorrect  == [][FlushBuffer => UNCHANGED RAF!vars]_vars
+SeekCorrect         == [][\A offset \in Offset: Seek(offset) => RAF!Seek(offset)]_vars
+SetLengthCorrect    == [][\A offset \in Offset: SetLength(offset) => RAF!SetLength(offset)]_vars
+SeekEstablishesInv2 == [][\A offset \in Offset: Seek(offset) => Inv2']_vars
+Write1Correct       == [][\A symbol \in Symbols: Write1(symbol) => RAF!Write(SeqToArray(<<symbol>>))]_vars
+Read1Correct        == [][\A symbol \in Symbols: Read1(symbol) => RAF!Read(SeqToArray(<<symbol>>))]_vars
+WriteAtMostCorrect  == [][\A len \in 1..MaxOffset: \A data \in Array(Symbols, len): WriteAtMost(data) => \E written \in 1..len: RAF!Write(ArraySlice(data, 0, written))]_vars
+ReadCorrect         == [][\A len \in 1..MaxOffset: \A data \in Array(Symbols, len): Read(data) => RAF!Read(data)]_vars
+
+\* Inv2 is a precondition for many actions; it should always be possible to
+\* restore Inv2.
+Inv2CanAlwaysBeRestored ==
```

Comment:
I translated the prose comment into the following property:

```tla
Inv2CanAlwaysBeRestoredPart2 ==
    [][IF dirty THEN FlushBuffer => Inv2' ELSE Seek(curr) => Inv2']_vars
```

, which TLC says is violated by the following trace:

```tla
Finished computing initial states: 1 distinct state generated at 2024-06-03 15:53:42.
Action property Inv2CanAlwaysBeRestored is violated.
The behavior up to this point is:
1: <Initial predicate>
/\ BuffSz = 2
/\ MaxOffset = 4
/\ dirty = FALSE
/\ length = 0
/\ curr = 0
/\ lo = 0
/\ buff = [elems |-> <<ArbitrarySymbol, ArbitrarySymbol>>]
/\ diskPos = 0
/\ file_content = [elems |-> <<>>]
/\ file_pointer = 0
/\ abstract_contents = [elems |-> <<>>]
2: <WriteAtMost([elems |-> <<A, A>>]) line 193, col 5 to line 202, col 60 of module BufferedRandomAccessFile>
/\ BuffSz = 2
/\ MaxOffset = 4
/\ dirty = TRUE
/\ length = 2
/\ curr = 2
/\ lo = 0
/\ buff = [elems |-> <<A, A>>]
/\ diskPos = 0
/\ file_content = [elems |-> <<>>]
/\ file_pointer = 0
/\ abstract_contents = [elems |-> <<A, A>>]
3: <FlushBuffer line 109, col 5 to line 119, col 43 of module BufferedRandomAccessFile>
/\ BuffSz = 2
/\ MaxOffset = 4
/\ dirty = FALSE
/\ length = 2
/\ curr = 2
/\ lo = 0
/\ buff = [elems |-> <<A, A>>]
/\ diskPos = 2
/\ file_content = [elems |-> <<A, A>>]
/\ file_pointer = 2
/\ abstract_contents = [elems |-> <<A, A>>]
```

Revisiting https://github.com/tlaplus/tlaplus/pull/907/files#r1625145343, this property, perhaps, too strong.  This hypothesis is supported because the following holds:

```tla
Inv2CanAlwaysBeRestoredPart2 ==
    [][Seek(curr) => Inv2']_vars
```

Reply:
This is a fantastic point. I will look into strengthening this property on Monday. Perhaps it can be fully specified using `\cdot` to compose `FillBuffer` and `Seek(curr)`.

Reply:
+1, I also forgot to mention that I would advise against using `\cdot` (gut feeling).